doc/mgmt/gnmi/SONiC_GNMI_Server_Interface_Design.md

```diff
@@ -0,0 +1,564 @@
+- [3 Serviceability and Debug](#3-serviceability-and-debug)
+- [4 Unit Tests](#4-unit-tests)
+  * [4.1 Unit Tests for Capabilities RPC](#41-unit-tests-for-capabilities-rpc)
+  * [4.2 Unit Tests for Get RPC](#42-unit-tests-for-get-rpc)
+  * [4.3 Unit Tests for Set RPC](#43-unit-tests-for-set-rpc)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+[Table 2: Mapping Operations](#table-2-mapping-operations)
+
+[Table 3: Translate Example](#table-3-translate-example)
+
+# Revision
+
+| Rev | Date        | Author             | Change Description  |
+|:---:|:-----------:|:------------------:|---------------------|
+| 0.1  | 05/05/2022 | Gang Lv    | Initial version     |
+
+# About this Document
+This document provides a detailed description on the strategy to implement the SONiC GNMI Server Interface.
+
+# Scope
+This document describes the high level design of SONiC GNMI Server Interface. 
+
+This document provides minor implementation details about the proposed solutions.
+
+# Definition/Abbreviation
+
+### Table 1: Abbreviations
+| **Term** | **Meaning**                |
+| -------- | -------------------------- |
+| gRPC     | A modern open-source high performance Remote Procedure Call (RPC) framework that can run in any environment |
+| gNMI     | gRPC Network Management Interface, used to retrieve or manipulate the state of a device via telemetry or configuration data |
+| gNOI     | gRPC Network Operations Interface |
+| JSON     | JavaScript Object Notation |
+| Yang     | Yet Another Next Generation, it is a modular language representing data structures in an XML tree format |
+| Generic Config Updater | SONiC configuration generic update and rollback feature |
+| ConfigDB | Configuration Database |
+| ApplDB | Application Database |
+
+# 1 Feature Overview
+
+SONiC should provide a gNMI/gNOI server interface where the client can communicate with SONiC (server) through the interface using their respective gRPC libraries based on ConfigDB/ApplDB/StateDB/CountersDB schema or SONiC Yang schema.
+
+To understand the requirements, we need to go over current SONiC management components:
+
+* SONIC CLI commands
+
+<img src="images/cli.svg" alt="cli" width="600px"/>
+
+1. Users could use SONiC CLI command "show runningconfig all" to read current configuration. 
+    - The blue arrow indicates this process.
+2. User could use SONiC CLI command "config apply-patch" to incrementally update configurations systematically and validate target configuration using SONiC Yang models. With generic_config_updater, all operations in a 'config apply-patch' command are processed in a single transaction that will either succeed or fail as a whole.
+    - "config apply-patch" supports ConfigDB, and Yang models.
+    - The green arrow indicates this process, and dotted arrow indicates Yang validation.
+3. SONiC CLI commands can use "config reload" to fully update configurations and validate target configuration using SONiC Yang models.
```

Comment:
>update [](http://example.com/codeflow?start=55&length=6)

To be more accurate:
fully update configurations -> erase original and fully reload configurations from a file